java-frontend/src/main/java/org/sonar/java/matcher/MethodMatcherFactory.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.matcher;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class MethodMatcherFactory {
+
+  private static final Pattern CLASS_PATTERN = Pattern.compile("^(\\w+[\\.\\w+]*(?:\\[\\])?)([\\(])?");
+  private static final Pattern METHOD_PATTERN = Pattern.compile("^(\\w+[\\.\\w+]*(?:\\[\\])?)#(\\w+)([\\(])?");
+  private static final Pattern ARGUMENT_PATTERN = Pattern.compile("(\\w+[\\.\\w+]*(?:\\[\\])?)([,\\)])?");
+
+  private MethodMatcherFactory() {
+    // no instances, only static, factory methods
+  }
+
+  public static MethodMatcher constructorMatcher(String descriptor) {
+    Matcher matcher = CLASS_PATTERN.matcher(descriptor);
+    if (!matcher.find()) {
+      throw new IllegalArgumentException("Illegal constructor specification: " + descriptor);
+    }
+    MethodMatcher constructorMatcher = MethodMatcher.create().typeDefinition(matcher.group(1)).name("<init>");
+    collectArguments(descriptor, matcher, 2, constructorMatcher);
+    return constructorMatcher;
+  }
+
+  public static MethodMatcher methodMatcher(String descriptor) {
+    Matcher matcher = METHOD_PATTERN.matcher(descriptor);
+    if (!matcher.find()) {
+      throw new IllegalArgumentException("Illegal method specification: " + descriptor);
+    }
+    MethodMatcher methodMatcher = MethodMatcher.create().typeDefinition(matcher.group(1)).name(matcher.group(2));
+    collectArguments(descriptor, matcher, 3, methodMatcher);
+    return methodMatcher;
+  }
+
+  public static void collectArguments(String descriptor, Matcher initialMatcher, int groupOffset, MethodMatcher methodMatcher) {
+    if ("(".equals(initialMatcher.group(groupOffset))) {
+      String remainder = descriptor.substring(initialMatcher.group().length());
+      if (!")".equals(remainder)) {
+        System.out.println(remainder);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Replace this usage of System.out or System.err by a logger. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS106)